crates/core/tedge_actors/src/servers/message_boxes.rs

```diff
@@ -123,3 +145,165 @@ impl<Request: Message, Response: Message> ClientMessageBox<Request, Response> {
+        // yet
+        // TODO: remove waits
+        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+        assert_eq!(rx.try_recv(), Err(TryRecvError::Empty));
+
+        // broadcast the first 4 tasks to finish
+        broadcast.send(()).unwrap();
+
+        // expect 5th task started executing only after the first 4 are complete
+        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+        assert_eq!(rx.recv().await, Some((4usize, 4)));
+    }
+
+    // The purpose of the test is to check that the server which uses a
+    // ConcurrentServerMessageBox terminates in a reasonable timeframe after
+    // receiving a shutdown request from the runtime. For this purpose we create
+    // a server which never completes its requests, and after filling it with
+    // requests up to its max concurrency level, we terminate the runtime.
+    #[tokio::test]
+    async fn does_not_block_runtime_exit() {
+        #[derive(Clone)]
+        struct TestServer;
+
+        #[async_trait]
+        impl Server for TestServer {
+            type Request = i32;
+            type Response = i32;
+
+            fn name(&self) -> &str {
+                ""
+            }
+
+            async fn handle(&mut self, _request: Self::Request) -> Self::Response {
+                eprintln!("got request: {_request:?}");
+                std::future::pending().await
+            }
+        }
+
+        struct TestServerBuilder {
+            box_builder: ServerMessageBoxBuilder<i32, i32>,
+        }
+
+        impl TestServerBuilder {
+            fn new() -> Self {
+                let box_builder =
+                    ServerMessageBoxBuilder::new("ConcurrentServerMessageBoxTest", 16);
+
+                Self { box_builder }
+            }
+
+            fn client_box(&mut self) -> SimpleMessageBox<i32, i32> {
+                self.box_builder.new_client_box(NoConfig)
+            }
+        }
+
+        impl Builder<ConcurrentServerActor<TestServer>> for TestServerBuilder {
+            type Error = std::convert::Infallible;
+
+            fn try_build(self) -> Result<ConcurrentServerActor<TestServer>, Self::Error> {
+                let message_box: ServerMessageBox<i32, i32> = self.box_builder.build();
+                let message_box = ConcurrentServerMessageBox::new(4, message_box);
+                Ok(ConcurrentServerActor::new(TestServer, message_box))
+            }
+        }
+
+        impl RuntimeRequestSink for TestServerBuilder {
+            fn get_signal_sender(&self) -> DynSender<RuntimeRequest> {
+                self.box_builder.get_signal_sender()
+            }
+        }
+
+        let mut runtime = Runtime::try_new(None).await.unwrap();
+
+        let mut server_actor_builder = TestServerBuilder::new();
+        let mut client_box = server_actor_builder.client_box();
+
+        runtime.spawn(server_actor_builder).await.unwrap();
+
+        // TODO: remove waits
```

Comment:
Why do you think these waits are useful in the current setting? Is this to avoid sending the shutdown request even before the server started to handle requests? If so you can use a channel for the tasks to tell they have been launched (instead of `eprintln!("got request: {_request:?}");`)

Reply:
My bad, forgot to delete the `eprintln!`, and indeed here it is very trivial to replace the waits with a channel.